<!DOCTYPE html>
<html lang="en"><head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <style>
        :root {
            --accent-color: #FF4D4D;
        }
    </style>

    
    
    
    
    
    

    
    <title>Section 2: Sequencing Data Analysis</title>
    <meta name="description" content="by School of Life Sciences at SUSTech &#43; BGI-Shenzhen">
    <meta name="keywords" content='blog, gokarna, hugo, data science, biology, genomics'>

    <meta property="og:url" content="https://dbrg77.github.io/SUSTech-BIOS201/posts/sequencing-data-analysis/">
    <meta property="og:type" content="website">
    <meta property="og:title" content="Section 2: Sequencing Data Analysis">
    <meta property="og:description" content="by School of Life Sciences at SUSTech &#43; BGI-Shenzhen">
    <meta property="og:image" content="/images/profile_avatar.jpg">

    <meta name="twitter:card" content="summary_large_image">
    <meta name="twitter:title" content="Section 2: Sequencing Data Analysis">
    <meta name="twitter:description" content="by School of Life Sciences at SUSTech &#43; BGI-Shenzhen">
    <meta property="twitter:domain" content="https://dbrg77.github.io/SUSTech-BIOS201/posts/sequencing-data-analysis/">
    <meta property="twitter:url" content="https://dbrg77.github.io/SUSTech-BIOS201/posts/sequencing-data-analysis/">
    <meta name="twitter:image" content="/images/profile_avatar.jpg">

    
    <link rel="canonical" href="https://dbrg77.github.io/SUSTech-BIOS201/posts/sequencing-data-analysis/" />

    <link rel="stylesheet" type="text/css" href="https://dbrg77.github.io/SUSTech-BIOS201//css/normalize.min.css" media="print" onload="this.media='all'">
    <link rel="stylesheet" type="text/css" href="https://dbrg77.github.io/SUSTech-BIOS201//css/main.css">
    <link disabled id="dark-theme" rel="stylesheet" href="https://dbrg77.github.io/SUSTech-BIOS201//css/dark.css">

    <script src="https://dbrg77.github.io/SUSTech-BIOS201//js/svg-injector.min.js"></script>
    <script src="https://dbrg77.github.io/SUSTech-BIOS201//js/feather-icons.min.js"></script>
    <script src="https://dbrg77.github.io/SUSTech-BIOS201//js/main.js"></script>

    
    
        <!-- KaTeX -->
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.css" integrity="sha384-6LkG2wmY8FK9E0vU9OOr8UvLwsaqUg9SETfpq4uTCN1agNe8HRdE9ABlk+fVx6gZ" crossorigin="anonymous">
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/katex.min.js" integrity="sha384-31El76TwmbHj4rF9DyLsygbq6xoIobG0W+jqXim+a3dU9W53tdH3A/ngRPxOzzaB" crossorigin="anonymous"></script>
    <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.16/dist/contrib/auto-render.min.js" integrity="sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl" crossorigin="anonymous"
        onload="renderMathInElement(document.body);"></script>
    <script>
        document.addEventListener("DOMContentLoaded", function() {
            renderMathInElement(document.body, {
                delimiters: [
                    {left: "$$", right: "$$", display: true},
                    {left: "$", right: "$", display: false}
                ]
            });
        });
    </script>
  
    
</head>
<body>
        <script type="text/javascript">
            
            setThemeByUserPref();
        </script><header class="header">
    <nav class="header-nav">

        
        <div class="avatar">
            <a href="https://dbrg77.github.io/SUSTech-BIOS201/">
                <img src="https://dbrg77.github.io/SUSTech-BIOS201//images/profile_avatar.jpg" alt="avatar" />
            </a>
        </div>
        

        <div class="nav-title">
            <a class="nav-brand" href="https://dbrg77.github.io/SUSTech-BIOS201/">BIOS201 Genome, why we are different?</a>
        </div>

        <div class="nav-links">
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIOS201"><span data-feather='home'></span> Home </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIOS201/posts/"><span data-feather='list'></span> Posts </a>
            </div>
            
            <div class="nav-link">
                <a href="https://dbrg77.github.io/SUSTech-BIOS201/about/"><span data-feather='user'></span> About </a>
            </div>
            
            <div class="nav-link">
                <a href="https://github.com/dbrg77/SUSTech-BIOS201"><span data-feather='github'></span>  </a>
            </div>
            

            <span class="nav-icons-divider"></span>
            <div class="nav-link dark-theme-toggle">
                <a>
                    <span id="theme-toggle-icon" data-feather="moon"></span>
                </a>
            </div>

            <div class="nav-link" id="hamburger-menu-toggle">
                <a>
                    <span data-feather="menu"></span>
                </a>
            </div>

            
            <ul class="nav-hamburger-list visibility-hidden">
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIOS201"><span data-feather='home'></span> Home </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIOS201/posts/"><span data-feather='list'></span> Posts </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://dbrg77.github.io/SUSTech-BIOS201/about/"><span data-feather='user'></span> About </a>
                </li>
                
                <li class="nav-item">
                    <a href="https://github.com/dbrg77/SUSTech-BIOS201"><span data-feather='github'></span>  </a>
                </li>
                
                <li class="nav-item dark-theme-toggle">
                    <a>
                        <span id="theme-toggle-icon" data-feather="moon"></span>
                    </a>
                </li>
            </ul>

        </div>
    </nav>
</header>
<main id="content">
    <div class="post container">
    <div class="post-header-section">
        <h1>Section 2: Sequencing Data Analysis</h1>
        <small role="doc-subtitle"></small>
        <p class="post-date">
            June 18, 2022
        </p>

        <ul class="post-tags">
        
        </ul>
    </div>

    <div class="post-content">
        <p>
            <p>Now that you have the sequencing data of yourself, let&rsquo;s identify the variants in your genome compared to the reference genome. In this section, we will only focus on single nucleotide polymorphisms (SNPs). Once we did that, we could look up in the public databases and literatures to see if the variants have any effect.</p>
<h2 id="21-computing-environment-setup">2.1 Computing environment setup</h2>
<p>The first step of doing anything is always to set up our computing environment, which involes in installing the software tools we need for our purposes. There are quite a few tools that can do what we want. Here, we just choose the following ones:</p>
<ul>
<li><a href="https://github.com/lh3/bwa"><strong>bwa</strong></a></li>
<li><a href="https://www.htslib.org"><strong>samtools, bcftools and tabix</strong></a></li>
</ul>
<p>During the lecture, we talked about using the following sequence of actions to install softwares:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>./configure    <span style="color:#93a1a1;font-style:italic"># collect necessary information from your machine</span>
</span></span><span style="display:flex;"><span>make           <span style="color:#93a1a1;font-style:italic"># compile the source code to binary machine code</span>
</span></span><span style="display:flex;"><span>make install   <span style="color:#93a1a1;font-style:italic"># copy the programs to your $PATH</span>
</span></span></code></pre></div><p>However, this can be very difficult for beginners. Therefore, we are going to use an easier route, that is, using a package management system. <a href="https://docs.conda.io/en/latest/"><strong>Conda</strong></a> is one of the most popular package management systems and environment management systems across different platforms, such as Windows, MacOS and Linux. In this course, we are using a minimal and light version of it: <a href="https://docs.conda.io/en/latest/miniconda.html"><strong>Miniconda</strong></a>. Go to the <a href="https://docs.conda.io/en/latest/miniconda.html"><strong>download page</strong></a> and get the installer for your system. If you are a Windows user, it is better to get the <a href="https://docs.microsoft.com/en-us/windows/wsl/install"><strong>Windows Subsystem for Linux (WSL)</strong></a> and install the Linux version of Miniconda uder the WSL. If you are a Mac user, it is better to install the command line tools by typing <code>xcode-select --install</code> in your terminal, and then download the bash version of the miniconda installer.</p>
<p>Then, you can simply install Miniconda by (using the Linux installer as an example):</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>bash Miniconda3-latest-Linux-x86_64.sh
</span></span></code></pre></div><p>Just accept the default settings, and you should not have any problems with the installation. Once that is done, we can create an <a href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html"><strong>environment</strong></a> by simply typing <code>conda create -n BIOS201</code> in your terminal. Then, we could activate the environment by <code>conda activate BIOS201</code>. If this is successful, you should see the string <code>(BIOS201)</code> in front of your prompt, like this:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>(BIOS201) xichen@DESKTOP-QAKDLCA:~$
</span></span></code></pre></div><p>Now, install the tools we need by simply type the following in the terminal:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>conda install -c bioconda bwa samtools bcftools tabix
</span></span></code></pre></div><p>Those tools, together with some commonly used commands in Linux, are all we need for this section.</p>
<h2 id="22-read-alignment-and-variant-calling">2.2 Read alignment and variant calling</h2>
<p>Once the required softwares are installed, you are ready to align your reads to the reference genome. First, create a directory called <code>bios201</code> by typing <code>mkdir bios201</code>. You can do everything inside that directory. Then, get the reference <code>fasta</code> file <a href="https://dbrg77.github.io/SUSTech-BIOS201/data/MGI358.SNP.fa.zip"><strong>MGI358.SNP.fa.zip</strong></a> (password protected due to administrative reasons), and put it in your directory. Unzip it by <code>unzip MGI358.SNP.fa.zip</code>. Then, build the index of the reference by typing the following command, which is required for the read alignment step later:</p>
<pre tabindex="0"><code>bwa index MGI358.SNP.fa
</code></pre><p>Once it is done, you should have some new files in the directory. Your directory should look like this:</p>
<pre tabindex="0"><code>bios201
├── MGI358.SNP.fa
├── MGI358.SNP.fa.amb
├── MGI358.SNP.fa.ann
├── MGI358.SNP.fa.bwt
├── MGI358.SNP.fa.pac
└── MGI358.SNP.fa.sa
</code></pre><p>Now you can put your own sequencing reads (the <code>.fastq.gz</code> file) to the <code>bios201</code> directory and start the read alignment. If your experiment has failed and you do not have your <code>fastq</code> file, you can just use one of the following data from the <a href="https://www.internationalgenome.org"><strong>1000 Genome Project</strong></a>:</p>
<ul>
<li><a href="https://dbrg77.github.io/SUSTech-BIOS201/data/HG00303_EUR.fastq.gz"><strong>HG00303_EUR.fastq.gz</strong></a></li>
<li><a href="https://dbrg77.github.io/SUSTech-BIOS201/data/HG00512_EAS.fastq.gz"><strong>HG00512_EAS.fastq.gz</strong></a></li>
<li><a href="https://dbrg77.github.io/SUSTech-BIOS201/data/HG01452_AMR.fastq.gz"><strong>HG01452_AMR.fastq.gz</strong></a></li>
<li><a href="https://dbrg77.github.io/SUSTech-BIOS201/data/HG01590_SAS.fastq.gz"><strong>HG01590_SAS.fastq.gz</strong></a></li>
<li><a href="https://dbrg77.github.io/SUSTech-BIOS201/data/HG03383_AFR.fastq.gz"><strong>HG03383_AFR.fastq.gz</strong></a></li>
</ul>
<p>Those data are truncated to just contain a few thousand reads around the locations we are intested in. It is small so that your laptop can handle them without any problems. I will use <code>HG00303_EUR.fastq.gz</code> for the demonstration. Align the reads using the following steps:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># align the reads to the reference</span>
</span></span><span style="display:flex;"><span>bwa mem -M -Y -t <span style="color:#2aa198;font-weight:bold">2</span> MGI358.SNP.fa HG00303_EUR.fastq.gz &gt; HG00303_EUR_to_MGI.sam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># filter and convert sam to bam</span>
</span></span><span style="display:flex;"><span>samtools view -SbF <span style="color:#2aa198;font-weight:bold">3844</span> HG00303_EUR_to_MGI.sam &gt; HG00303_EUR_to_MGI.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># sort the bam file</span>
</span></span><span style="display:flex;"><span>samtools sort HG00303_EUR_to_MGI.bam -T tmp -o HG00303_EUR_to_MGI_sorted.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># index the bam file</span>
</span></span><span style="display:flex;"><span>samtools index HG00303_EUR_to_MGI_sorted.bam
</span></span></code></pre></div><p>If you are familiar with Unix command lines, those steps can be &ldquo;piped&rdquo; together and done in fewer lines of code without creating those intermediate files:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># align, filter, convert and sort</span>
</span></span><span style="display:flex;"><span>bwa mem -M -Y -t <span style="color:#2aa198;font-weight:bold">2</span> MGI358.SNP.fa HG00303_EUR.fastq.gz | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    samtools view -SuF <span style="color:#2aa198;font-weight:bold">3844</span> - | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    samtools sort - -T tmp -o HG00303_EUR_to_MGI_sorted.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># index the bam file</span>
</span></span><span style="display:flex;"><span>samtools index HG00303_EUR_to_MGI_sorted.bam
</span></span></code></pre></div><p>Finally, we could identify SNPs between your DNA and the reference using bcftools:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>bcftools mpileup -f MGI358.SNP.fa HG00303_EUR_to_MGI_sorted.bam | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    bcftools call -mv - &gt; HG00303_EUR_to_MGI_all.vcf
</span></span></code></pre></div><p>The output is in a Variant Call Format (VCF). You can check the specification by <a href="https://samtools.github.io/hts-specs/VCFv4.2.pdf"><strong>clicking here</strong></a>. Your output file should look like this (I omitted a lot of header lines here):</p>
<pre tabindex="0"><code>##fileformat=VCFv4.2
##FILTER=&lt;ID=PASS,Description=&#34;All filters passed&#34;&gt;
##bcftoolsVersion=1.15.1+htslib-1.15.1
##bcftoolsCommand=mpileup -f MGI358.SNP.fa HG00303_EUR_to_MGI_sorted.bam
##reference=file://MGI358.SNP.fa
##contig=&lt;ID=rs1490413,length=2084&gt;
##contig=&lt;ID=rs5745448,length=2078&gt;
##contig=&lt;ID=rs3737576,length=2079&gt;
.
.
.
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  HG00303_EUR_to_MGI_sorted.bam
rs1490413       1000    .       C       T       10.7923 .       DP=1;SGB=-0.379885;FS=0;MQ0F=0;AC=2;AN=2;DP4=0,0,1,0;MQ=60      GT:PL   1/1:40,3,0
rs1490413       1024    .       G       A       225.417 .       DP=253;VDB=0;SGB=-0.693147;FS=0;MQ0F=0;AC=2;AN=2;DP4=0,0,253,0;MQ=59    GT:PL   1/1:255,255,0
</code></pre><p>Your experiments are done using a set of primers for targeted sequencing. It means you know what they are. Actually, the name of the record in the <code>MGI358.SNP.fa</code> file is the ID from the database <a href="https://www.ncbi.nlm.nih.gov/snp/"><strong>dbSNP</strong></a>. You can see that the first column of your VCF file is basically the SNP ID (<code>rsXXXX</code>). This allows you to simply search the ID in the dbSNP database to see if we already have some knowledge about them.</p>
<h3 id="practice">Practice:</h3>
<ul>
<li>Examine the first 2 records in your VCF file in the <a href="https://www.ncbi.nlm.nih.gov/snp/"><strong>dbSNP</strong></a> database.</li>
</ul>
<h2 id="23-whole-genome-sequencing-wgs">2.3 Whole Genome Sequencing (WGS)</h2>
<p>Another common scenario is that you are doing this type of experiment in a &ldquo;<em>de novo</em>&rdquo; way. You sequence your whole genome, try to identify all the SNPs in the genome and investigate if they have any effects. Now let&rsquo;s perform the analysis in this scenario.</p>
<p>In order to do this, we need the reference sequence of the entire human genome. You can downlaod the <code>fasta</code> file from the <a href="https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/"><strong>UCSC Genome Browser</strong></a>, but the file size is very big and the interenet speed at our university is slow. I already prepared them for you. You can download the <a href="http://10.20.42.238:1234/folders/sf_html/BIOS201/GRCh38.tar.gz"><strong>GRCh38.tar.gz</strong></a> file (file size: 4.1 GB) located on my workstation (campus access only) to your <code>bios201</code> directory. Extract the file by simply typing <code>tar zxvf GRCh38.tar.gz</code>. Once it is done, you will get a directory called <code>GRCh38</code> and inside it, there are seven files (a total of 8 GB):</p>
<pre tabindex="0"><code>GRCh38.fa
GRCh38.fa.amb
GRCh38.fa.ann
GRCh38.fa.bwt
GRCh38.fa.fai
GRCh38.fa.pac
GRCh38.fa.sa
</code></pre><p>The bwa index files are already there, so you don&rsquo;t need to index the reference <code>fasta</code> again. Now, let&rsquo;s pretend the <code>fastq</code> file you have is from the whole genome sequencing of yourself and map the reads to the whole genome and call variants. Basically, we are repreating the previous procedures using the whole genome as the reference:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># align, filter, convert and sort</span>
</span></span><span style="display:flex;"><span>bwa mem -M -Y -t <span style="color:#2aa198;font-weight:bold">2</span> GRCh38/GRCh38.fa HG00303_EUR.fastq.gz | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    samtools view -SuF <span style="color:#2aa198;font-weight:bold">3844</span> - | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    samtools sort - -T -o HG00303_EUR_to_GRCh38_sorted.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># index the bam file</span>
</span></span><span style="display:flex;"><span>samtools index HG00303_EUR_to_GRCh38_sorted.bam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># call variants</span>
</span></span><span style="display:flex;"><span>bcftools mpileup -f GRCh38/GRCh38.fa HG00303_EUR_to_GRCh38_sorted.bam | <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>    bcftools call -mv - &gt; HG00303_EUR_to_GRCh38_all.vcf
</span></span></code></pre></div><p>Now the new VCF file <code>HG00303_EUR_to_GRCh38_all.vcf</code> looks like this:</p>
<pre tabindex="0"><code>##fileformat=VCFv4.2
##FILTER=&lt;ID=PASS,Description=&#34;All filters passed&#34;&gt;
##bcftoolsVersion=1.9+htslib-1.9
##bcftoolsCommand=mpileup -f GRCh38/GRCh38.fa HG00303_EUR_to_GRCh38_sorted.bam
##reference=file://GRCh38/GRCh38.fa
##contig=&lt;ID=chr1,length=248956422&gt;
##contig=&lt;ID=chr10,length=133797422&gt;
##contig=&lt;ID=chr11,length=135086622&gt;
##contig=&lt;ID=chr12,length=133275309&gt;
##contig=&lt;ID=chr13,length=114364328&gt;
##contig=&lt;ID=chr14,length=107043718&gt;
##contig=&lt;ID=chr15,length=101991189&gt;
##contig=&lt;ID=chr16,length=90338345&gt;
##contig=&lt;ID=chr17,length=83257441&gt;
##contig=&lt;ID=chr18,length=80373285&gt;
##contig=&lt;ID=chr19,length=58617616&gt;
##contig=&lt;ID=chr2,length=242193529&gt;
##contig=&lt;ID=chr20,length=64444167&gt;
##contig=&lt;ID=chr21,length=46709983&gt;
##contig=&lt;ID=chr22,length=50818468&gt;
##contig=&lt;ID=chr3,length=198295559&gt;
##contig=&lt;ID=chr4,length=190214555&gt;
##contig=&lt;ID=chr5,length=181538259&gt;
##contig=&lt;ID=chr6,length=170805979&gt;
##contig=&lt;ID=chr7,length=159345973&gt;
##contig=&lt;ID=chr8,length=145138636&gt;
##contig=&lt;ID=chr9,length=138394717&gt;
##contig=&lt;ID=chrM,length=16569&gt;
##contig=&lt;ID=chrX,length=156040895&gt;
##contig=&lt;ID=chrY,length=57227415&gt;
.
.
.
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  HG00303_EUR_to_GRCh38_sorted.bam
chr1    76344917        .       A       C       74      .       DP=3;VDB=0.431015;SGB=-0.511536;MQSB=1;MQ0F=0;AC=2;AN=2;DP4=0,0,2,1;MQ=60       GT:PL   1/1:104,9,0
chr1    196145102       .       G       A       132     .       DP=6;VDB=0.0602347;SGB=-0.590765;MQSB=1;MQ0F=0;AC=2;AN=2;DP4=0,0,4,1;MQ=60      GT:PL   1/1:162,15,0
</code></pre><p>The content in the current VCF file is slightly different from the previous one. The first column is now the chromosome names. The second column indicates the position of the SNP. Those are useful and important pieces of information. However, do the variants have any effect on our phenotypes? Do we know anything about them? In order to figue out that, we need to compare the variants to a database containing many SNPs that we know about. If the variant in your genome is present in the database, it is relative easy to find out what it does based on literatures.</p>
<p>You probably noticed that the values in the third column of the current VCF file are all <code>.</code>. Now let&rsquo;s annotate the variant with the <code>rsXXXX</code> ID in the dbSNP database. First, we need the variants in the database in the VCF format. Go to the <a href="https://www.ncbi.nlm.nih.gov/snp/"><strong>dbSNP database</strong></a>, then click the <code>FTP Download</code> at the right hand side. Go all the way to the <code>organisms -&gt; human_9606_b151_GRCh38p7 -&gt; VCF</code> directory, and the <code>00-All.vcf.gz</code> file is what you need. However, it is really big, and I have prepared a subset that is enough for us to use in this section. You can download the truncated version (<a href="https://dbrg77.github.io/SUSTech-BIOS201/data/dbSNP_subset_formatted.vcf"><strong>dbSNP_subset_formatted.vcf</strong></a>) to your own compouter and put it under the <code>bios201</code> directory.</p>
<p>Before we could perform the annotation, we need to compress and index the VCF files using <code>bgzip</code> and <code>tabix</code>:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># compress with bgzip</span>
</span></span><span style="display:flex;"><span>bgzip HG00303_EUR_to_GRCh38_all.vcf
</span></span><span style="display:flex;"><span>bgzip dbSNP_subset_formatted.vcf
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#93a1a1;font-style:italic"># index the file</span>
</span></span><span style="display:flex;"><span>tabix -s <span style="color:#2aa198;font-weight:bold">1</span> -b <span style="color:#2aa198;font-weight:bold">2</span> -e <span style="color:#2aa198;font-weight:bold">2</span> -p vcf HG00303_EUR_to_GRCh38_all.vcf.gz
</span></span><span style="display:flex;"><span>tabix -s <span style="color:#2aa198;font-weight:bold">1</span> -b <span style="color:#2aa198;font-weight:bold">2</span> -e <span style="color:#2aa198;font-weight:bold">2</span> -p vcf dbSNP_subset_formatted.vcf.gz
</span></span></code></pre></div><p>You will see the VCF files are compressed into the <code>.gz</code> files and for each of them, there is a <code>.tbi</code> index file. Now, you are ready to annotate your own VCF using bcftools:</p>
<div class="highlight"><pre tabindex="0" style="color:#586e75;background-color:#eee8d5;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-bash" data-lang="bash"><span style="display:flex;"><span>bcftools annotate -c ID -O v <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>                  -a dbSNP_subset_formatted.vcf.gz <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>                  -o HG00303_EUR_to_GRCh38_all_annotated.vcf <span style="color:#2aa198">\
</span></span></span><span style="display:flex;"><span><span style="color:#2aa198"></span>                  HG00303_EUR_to_GRCh38_all.vcf.gz
</span></span></code></pre></div><p>The program will look at the variants in your own VCF and the dbSNP database VCF files and compare them to see if they are the same. If yes, then the program will replace the <code>.</code> in the third column of your VCF with the ID from the dbSNP VCF (<em>i.e.</em> the <code>rsXXXX</code> ID). The advantage is that those IDs are unique and everybody is using them. You can easily communicate with others and look at the potential effect in the database.</p>
<p>Now look inside your annoated VCF file, do you notice something different? You probably see that the third column of a lot of the records is replaced by the <code>rsXXXX</code> ID, while some of them still remain as <code>.</code>, which means it does not match any record in the dbSNP database. There are many reasons for this, and you can think about this by yourself.</p>
<h3 id="practice-1">Practice</h3>
<ul>
<li>Find the SNP <code>rs1821380</code> in your annotated VCF file, can you find it or not? If yes, what does the record tell you? If not, can you think of possible reasons?</li>
<li>Look it up in the dbSNP database to see if we know anything about this variant</li>
<li>What are the frequencies of the reference and alternative alleles, respectively</li>
</ul>
<h2 id="24-visualisation-of-the-read-alignment">2.4 Visualisation of the read alignment</h2>
<p>Very often, the most efficient way of getting a sense of your data is the least scientific way, that is, looking at it by eyes. To visualise our sequencing reads in the genome, you can use <a href="https://igv.org"><strong>IGV: Integrative Genomics Viewer</strong></a>. Go to the <a href="https://software.broadinstitute.org/software/igv/download"><strong>Download page</strong></a> and get the program (with Java included) for your system. Lauch the program and load the <code>bam</code> file generated in the previous step. As long as the bam index (the <code>bai</code> file) is located in the same directory of your <code>bam</code> file, it should be loaded without any problem.</p>
<h3 id="practice-2">Practice</h3>
<p>Look at the alignments around the following SNPs and see if your DNA is different from the reference genome:</p>
<ul>
<li>rs1821380 (chr15:39021201)</li>
<li>rs1800414 (chr15:27951891)</li>
<li>rs671 (chr12:111803962)</li>
<li>rs885479 (chr16:89919746)</li>
</ul>

        </p>
    </div>
</div>



    

        </main><footer class="footer">
    <span>&copy; 2022 SUSTech BIOS201</span>
    <span>
        Made with &#10084;&#65039; using <a target="_blank" href="https://github.com/526avijitgupta/gokarna">Gokarna</a>
    </span>
</footer>
</body>
</html>
